opensearch/src/main/java/org/opensearch/sql/opensearch/data/value/OpenSearchExprValueFactory.java

```diff
@@ -348,21 +364,37 @@ private ExprValue parseStruct(Content content, String prefix) {
   }
 
   /**
-   * Todo. ARRAY is not completely supported now. In OpenSearch, there is no dedicated array type.
-   * <a href="https://opensearch.org/docs/latest/opensearch/supported-field-types/nested/">docs</a>
-   * The similar data type is nested, but it can only allow a list of objects.
+   * Parse array content. Can also parse nested which isn't necessarily an array.
+   * @param content Content to parse.
+   * @param prefix Prefix for Level of object depth to parse.
+   * @param type Type of content parsing.
+   * @return Value parsed from content.
    */
-  private ExprValue parseArray(Content content, String prefix) {
+  private ExprValue parseArray(Content content, String prefix, ExprType type) {
     List<ExprValue> result = new ArrayList<>();
-    // ExprCoreType.ARRAY does not indicate inner elements type.
-    if (Iterators.size(content.array()) == 1 && content.objectValue() instanceof JsonNode) {
-      result.add(parse(content, prefix, Optional.of(STRUCT)));
+
+    // ARRAY is mapped to nested but can take the json structure of an Object
+    if (content.objectValue() instanceof ObjectNode) {
+      result.add(parseStruct(content, prefix));
     } else {
       content.array().forEachRemaining(v -> {
-        // ExprCoreType.ARRAY does not indicate inner elements type. OpenSearch nested will be an
-        // array of structs, otherwise parseArray currently only supports array of strings.
-        if (v.isString()) {
+        if (type instanceof OpenSearchIpType
```

Comment:
can we do `type instanceof OpenSearchDataType`?

Reply:
Sadly no!